Some API fixes discovered while experimenting with OpenAPI Generator generated client library

gohrner opened the pull request:
While playing with the OpenProject API spec and the OpenAPI Generator for the last one or two months, I stumbled across quite a few problems with the API spec.

Some of the issues meanwhile have already been ironed out upstream - in these cases I was too slow to push my adjustments - but several of the problems still persist.

Please show some patience, as I'm not yet very experienced with OpenAPI V3.

See also: https://community.openproject.org/topics/14323

Commit:
Try to iron out some peculiarities in APIv3 spec yml files.

Commit:
work_package_model: adjust Collection types

Attempt to adjust Collection types as documented in the description.

Commit:
Fixed response schema of /api/v3/work_packages/{id}/activities

Kharonus commented on docs/api/apiv3/paths/work_packages.yml:
```diff
@@ -424,6 +424,7 @@ post:
       application/json:
         schema:
+          "$ref": "../components/schemas/work_package_model.yml"
           example:
```
you can remove the example here, too. There is an example defined in `work_package_model.yml`, and due to the `$ref` any siblings are ignored nevertheless.

Kharonus commented on docs/api/apiv3/paths/work_package.yml:
```diff
@@ -484,24 +484,11 @@ patch:
       application/json:
         schema:
+          "$ref": "../components/schemas/work_package_model.yml"
           example:
```
same here, please remove example.

Kharonus commented on docs/api/apiv3/paths/work_package_activities.yml:
```diff
@@ -57,7 +57,14 @@ get:
                 count: 2
                 total: 2
           schema:
-            "$ref": "../components/schemas/work_package_activities_model.yml"
```
nesting the ref in a collection here is wrong, as the `work_package_activities_model.yml` is already meant to be a collection. (Even if it is currently still an object without schema, only with example.) Please revert.

Kharonus commented on docs/api/apiv3/components/schemas/link.yml:
```diff
@@ -20,7 +20,7 @@ properties:
     description: The HTTP verb to use when requesting the resource
   payload:
-    type: string
+    type: object
```
Can you provide an example endpoint, where we actually do return an object here in payload?

gohrner replied:
The "problem" is that in many cases, structured JSON objects are contained, not "naked" JSON strings, which would be what's expected with type "string".

One example taken from https://www.openproject.org/docs/api/endpoints/work-packages/#create-work-package :


```json
"payload": {
   "user": {
      "href": "/api/v3/users/{user_id}"
   }
},
```

Kharonus replied:
you're absolutely right. checked that code in production, too. payload must be `object`

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -221,24 +224,29 @@ properties:
-          readOnly: true
+              **Permission** view work packages
+            readOnly: true
       attachments:
```
`attachments` is indeed no collection of links, but just a single link to the attachments collection, please revert

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -382,17 +393,20 @@ properties:
 
             **Resource**: Version
       watchers:
```
`watchers` is no collection of links, but a single link that points to a collection ressource, please revert.

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -285,17 +293,20 @@ properties:
+          allOf:
+          - "$ref": "./link.yml"
+          - description: |-
+              Array of all visible children of the work package
```
the description no longer matches. It should instead be s.th. like 

`A visible child work package of the current work package.`

conditions are the same, resource is `WorkPackage`

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -221,24 +224,29 @@ properties:
+          allOf:
+          - "$ref": "./link.yml"
+          - description: |-
+              Array of all visible ancestors of the work package, with the root node being the first element
```
the description no longer matches. It should instead be s.th. like

A visible ancestor work package of the current work package.

conditions are the same, resource is WorkPackage

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -143,11 +143,14 @@ properties:
+        items:
+          allOf:
+          - "$ref": "./link.yml"
+          - description: Collection of predefined changes that can be applied to the
```
the description no longer matches. It should instead be s.th. like

A predefined action that can be applied to the work package.

**Resource**: CustomAction

Kharonus reviewed (changes requested):
Many thanks again for your dedicated work. I have added some comments and requested some changes.

Commit:
Merge branch 'dev' into api_spec_fixes_1

Manually Merged Conflicts:
	docs/api/apiv3/components/schemas/work_package_model.yml

Commit:
Incorporated review feedback from Kharonus.

Thanks for the review!

Mainly superfluous examples, mismatching documentation and wrong
adjustments on my end.

See PR https://github.com/opf/openproject/pull/10175 for details.

Kharonus commented on docs/api/apiv3/components/schemas/work_package_model.yml:
```diff
@@ -143,10 +143,15 @@ properties:
+        readOnly: true
+        items:
+          allOf:
           - "$ref": "./link.yml"
```
Please check your formatter for `yaml` files. While it is not needed for yaml syntax, I try to harmonize most of our yaml files to one style. and within this style, a list should be indented by 2 spaces from its parent. Applying here on the `allOf` objects.

gohrner replied:
Actually all the files were manually formatted, as my yaml editor simply messes up the existing formatting. :-/

Which editor do you use / can you recommend?

Anyway, formatting should be better now (I hope).

Kharonus replied:
I can recommend the JetBrains tools, but they are not OpenSource or free to use. They do have a some subscriptions for free, e.g. if you are a student, teacher or working for non-commercial OpenSource projects.

This should not be some advertising, hence I won't link to their pricing page in one of our public PRs. ;) But as you asked me, what I can recommend, I find their IDEs the most intuitive and powerful ones. And they have in-built formatting support for all relevant formats, which can read `.editorconfig` and overwrite it locally, too.

If you want to go really specialized, you can of course run VS Code (MIT-License, completely free and OpenSource), and configure it to the right job for you. This editor is one of the most popular, for good reason.

Kharonus commented on docs/api/apiv3/paths/work_packages.yml:
```diff
@@ -435,48 +435,5 @@ post:
-            subject:
-              type: string
-          type: object
+          "$ref": "../components/schemas/work_package_model.yml"
```
can you replace the response body with the `$ref`, too, please?

Just, FYI: I started on file links with introducing a separation of read and write models. This will eventually take place here, too. But for now, we just reference the `work_package_model.yml` for the request and the response body.

gohrner replied:
Mh, not quite sure what you're referring to - all schemas the `responses` section already are `$ref`s, as far as I can tell. (There are verbose `examples` included there, are you referring to these?)

I also could not find a `responseBody` in the OpenAPI documentation, in contrast to `requestBody`.

Kharonus replied:
you are right, I was apparently too tired that day, to see, that the whole clutter is just the example. response bodies in OpenAPI are what is defined in https://swagger.io/specification/#response-object > content/[mime-type]/schema.

Kharonus reviewed (changes requested):
Formatting and a small improvement plea.

Commit:
Unified formatting.

https://github.com/opf/openproject/pull/10175#discussion_r832348434

Kharonus reviewed (approved):
LGTM
